conan/cli/commands/cache.py

```diff
@@ -31,30 +33,34 @@ def cache_path(conan_api: ConanAPI, parser, subparser, *args):
     args = parser.parse_args(*args)
     try:
         pref = PkgReference.loads(args.reference)
+        ref = None
     except ConanException:
         pref = None
-
-    if not pref:  # Not a package reference
         ref = RecipeReference.loads(args.reference)
-        if args.folder is None:
-            path = conan_api.cache.export_path(ref)
-        elif args.folder == "export_source":
-            path = conan_api.cache.export_source_path(ref)
-        elif args.folder == "source":
-            path = conan_api.cache.source_path(ref)
-        elif args.folder == "metadata":
-            path = conan_api.cache.recipe_metadata_path(ref)
-        else:
-            raise ConanException(f"'--folder {args.folder}' requires a valid package reference")
-    else:
-        if args.folder is None:
-            path = conan_api.cache.package_path(pref)
-        elif args.folder == "build":
-            path = conan_api.cache.build_path(pref)
-        elif args.folder == "metadata":
-            path = conan_api.cache.package_metadata_path(pref)
-        else:
-            raise ConanException(f"'--folder {args.folder}' requires a recipe reference")
+
+    # Checking exclusive folders depending on the reference
+    if args.folder in ['export_source', 'source'] and ref is None:
+        raise ConanException(f"'--folder {args.folder}' requires a recipe reference")
+    elif args.folder == 'build' and pref is None:
+        raise ConanException(f"'--folder {args.folder}' requires a package reference")
+    elif args.folder is None:
+        # FIXME: if the package_path(pref) does not exist, this function is raising an error
+        #        (actually, _resolve_latest_pref() does).
+        #        Should build_path, package_metadata_path, etc. do the same?
+        path = conan_api.cache.export_path(ref) if ref else conan_api.cache.package_path(pref)
+    elif args.folder == "export_source":
+        path = conan_api.cache.export_source_path(ref)
+    elif args.folder == "source":
+        path = conan_api.cache.source_path(ref)
+    elif args.folder == "metadata":
+        path = conan_api.cache.recipe_metadata_path(ref) if ref else \
+               conan_api.cache.package_metadata_path(pref)
+    elif args.folder == "build":
+        path = conan_api.cache.build_path(pref)
+    # Assuming that pref or ref would have raised an exception if args.folder is None
+    if args.folder and not os.path.exists(path):
+        reference = f'recipe reference {ref}' if ref else f'package reference {pref}'
```

Comment:
Better not change this, this looks like a potential refactor, but it hides the actual behavior change of raising the errors when folders do not exist.